corehq/project_limits/gauge.py

```diff
@@ -0,0 +1,188 @@
+        :param timeout: ``int`` how many seconds the reported value should remain in cache.
+        """
+        self.feature_key = feature_key
+        self.timeout = timeout
+        self.scopes = scope_fn(feature_key)
+
+    def report(self, tracked_key, observerd_value):
+        """
+        :param tracked_key: ``str`` A key that identifies the tracked event
+        :param observerd_value: ``float`` The value of the tracked event
+        Saves the observed value for the tracked key and also updates last reported time.
+        """
+        assert tracked_key in self.scopes, f"{tracked_key} not in the following scopes {self.scopes}"
+
+        cache.set(self._tracked_event_cache_key(tracked_key), observerd_value, timeout=self.timeout)
+
+        cache.set(self._last_reported_timestamp_key, datetime.now(tz=timezone.utc), timeout=self.timeout)
+
+    @quickcache(['self.feature_key'], timeout=5 * 60, memoize_timeout=60)
+    def get_values(self):
+        """
+        returns a list of all the reported values for
+        all the scopes for the give feature key.
+        """
+        all_cache_keys = [self._tracked_event_cache_key(key) for key in self.scopes]
+        all_observed_values = cache.get_many(all_cache_keys)
+        return list(all_observed_values.values())
+
+    def _tracked_event_cache_key(self, tracked_key):
+        return f'{type(self).__name__}-{tracked_key}-key'
+
+    @property
+    def _last_reported_timestamp_key(self):
+        return f'{type(self).__name__}-{self.feature_key}-last-reported'
+
+    def get_last_reported_time(self):
+        "returns the last time when the lag was reported or will return ``datetime.min`` if not reported yet"
+        return cache.get(self._last_reported_timestamp_key, datetime.min)
+
+    def max(self):
+        """
+        return max lag for a given topic
+        """
+        values = self.get_values()
+        return max(values) if values else None
+
+    def min(self):
+        """
+        return min lag for a given topic
+        """
+        values = self.get_values()
+        return min(values) if values else None
+
+    def avg(self):
+        """
+        return average lag for a given topic
+        """
+        values = self.get_values()
+        return sum(values) / len(values) if values else None
+
+
+case_pillow_lag_gauge = Gauge(feature_key=CASE_SQL, scope_fn=get_all_kafka_partitons_for_topic)
+
+
+class GaugeLimiter:
+
+    def __init__(self, gauge, get_gauge_definition):
+        """
+        :param gauge: ``Gauge`` An instance of Gauge class
+        :param get_gauge_defintions: ``function``
+            A function to get config that GaugeLimiter will use to
+            figure out throttling conditions.
+        """
+        self.gauge = gauge
+        self.gauge_definition_fn = get_gauge_definition
+
+    def wait(self):
+        if not self.allow_usage():
+            time.sleep(self.gauge_definition.throttle_for_seconds)
```

Comment:
```suggestion
            time.sleep(self.gauge_definition.wait_for_seconds)
```

Reply:
Ahh! Good catch. Missed this one.

Reply:
https://github.com/dimagi/commcare-hq/pull/34800/commits/d980bd2fa35b0cfa4d635ae69bf8c2a4c3813392